pkg/receive/handler.go

```diff
@@ -747,34 +755,55 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
 	// This is needed if context is canceled or if we reached success of fail quorum faster.
 	defer func() {
 		go func() {
-			for err := range ec {
+			for errReplicaFn := range ec {
+				_, err := errReplicaFn()
 				if err != nil {
 					level.Debug(tLogger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
 				}
 			}
 		}()
 	}()
 
-	var success int
+	var (
+		// success counts how many requests per replica have succedeed.
+		success = make([]int, h.options.ReplicationFactor)
```

Comment:
This change initializes some slices. Maybe it would be a good idea to benchmark and see changes to hot path?

Reply:
I ran `BenchmarkHandlerReceiveHTTP`, I think we won't be able to avoid those few extra allocs but I think it should still be acceptable?

```
name                                                                                old time/op    new time/op    delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                         891µs ± 0%     942µs ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.07ms ± 0%    1.12ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       9.42ms ± 0%    9.68ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          11.5ms ± 0%    11.2ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      39.7ms ± 0%    39.3ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         49.7ms ± 0%    49.7ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                 85.7ms ± 0%   105.6ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12    87.6ms ± 0%    89.9ms ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old alloc/op   new alloc/op   delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                        1.15MB ± 0%    1.15MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.41MB ± 0%    1.41MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       13.1MB ± 0%    13.1MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          15.5MB ± 0%    15.5MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      53.4MB ± 0%    53.4MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         62.0MB ± 0%    62.0MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                  110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12     110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old allocs/op  new allocs/op  delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                         3.10k ± 0%     3.10k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12            6.63k ± 0%     6.64k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                        30.3k ± 0%     30.3k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12           65.2k ± 0%     65.2k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                        121k ± 0%      121k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12           260k ± 0%      260k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                   89.0 ± 0%      93.0 ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12       217 ± 0%       223 ± 0%   ~     (p=1.000 n=1+1)
```

Reply:
Yup! Thanks!

Reply:
With latest changes, still looking fine:

```
name                                                                                old time/op    new time/op    delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                        1.20ms ± 0%    0.87ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.25ms ± 0%    1.05ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       10.6ms ± 0%     9.3ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          12.8ms ± 0%    10.8ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      46.3ms ± 0%    36.8ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         55.5ms ± 0%    43.6ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                  102ms ± 0%      82ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12     101ms ± 0%      86ms ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old alloc/op   new alloc/op   delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                        1.15MB ± 0%    1.15MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.41MB ± 0%    1.41MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       13.1MB ± 0%    13.1MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          15.5MB ± 0%    15.5MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      53.3MB ± 0%    53.2MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         62.0MB ± 0%    62.0MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                  110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12     110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old allocs/op  new allocs/op  delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                         3.10k ± 0%     3.10k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12            6.63k ± 0%     6.64k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                        30.3k ± 0%     30.3k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12           65.2k ± 0%     65.2k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                        121k ± 0%      121k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12           260k ± 0%      260k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                   89.0 ± 0%      94.0 ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12       218 ± 0%       224 ± 0%   ~     (p=1.000 n=1+1)

```